### PR TITLE
Microsoft.Support RP: Swagger and examples for public preview

### PR DESCRIPTION
**Pull request by shahbj79**
1. Swagger for Microsoft.Support RP public preview version
2. Here is the [link ](https://github.com/Azure/azure-rest-api-specs-pr/pull/753/) to private repo PR
3. Here's a summary of differences when compared to private repo PR
   a. Added support for PUT API to create SupportTicket and removed CreateSupportTicket POST 
      operation.
   b.  SupportTicketDetails model updates
   c.  TechnicalTicketDetails and QuotaTicketDetails model updates
   d.  Example updates

### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/adx-documentation-pr/wiki/Overall-basic-flow) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [ ] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [ ] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

**Review by weidongxu-microsoft — COMMENTED** [review]
This appears to be a new resource provider. Please make sure necessary process is done. And please invite as many reviewer from the team as possible.

https://armwiki.azurewebsites.net/rp_onboarding/RPOnboardingOverview.html

I am not sure whether you want to start as a private reivew first, to limit exposure.

File: specification/support/resource-manager/readme.java.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+## Java
```
[the remark this answers is not in the document]
Comment (shahbj79) [reply specification/support/resource-manager/readme.java.md]: @weidongxu-microsoft - we are an existing RP, but we don't have any swagger doc for our very old 2015 version. This is a new version we are releasing which has gone through private preview process. I have mentioned this in the pull request description and also provided link to the private repo PR and summarized the delta from that PR.

[the remark this answers is not in the document]
Comment (shahbj79) [reply specification/support/resource-manager/readme.java.md]: @weidongxu-microsoft - I will take a look at the example that you provided. Thanks.

[the remark this answers is not in the document]
Comment (shahbj79) [reply specification/support/resource-manager/readme.java.md]: @weidongxu-microsoft - Can you also help me understand why "azure-sdk-for-net" task is failing? Not sure what I need to do to fix this. I am able to generate this locally.

File: specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json [hunk 2]
```diff
@@ -0,0 +1,1473 @@
...
+            "$ref": "#/definitions/Operation"
+          }
+        }
+      }
+    },
+    "Operation": {
+      "description": "The operation supported by Microsoft Support RP.",
+      "type": "object",
+      "properties": {
+        "name": {
```
[the remark this answers is not in the document]
Comment (shahbj79) [reply specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json]: Addressed.

File: specification/support/resource-manager/readme.java.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (shahbj79) [reply specification/support/resource-manager/readme.java.md]: updated readme.java.md

File: specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json [hunk 3]
```diff
@@ -0,0 +1,1487 @@
...
+                    "description": "Country of the user. This is the ISO Alpha-3 code",
+                    "type": "string"
+                },
+                "preferredSupportLanguage": {
+                    "description": "Language of the user. This is the standard country-language code.",
+                    "type": "string"
+                }
+            }
+        },
+        "UpdateContactProfile": {
```
Comment (KrisBash) [inline specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json]: Are these exposed in an existing api? Have they been cleared by privacy/GDPR?

**Review by KrisBash — CHANGES_REQUESTED** [review]
Please see comments

File: specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json (hunk 3 above)
Comment (shahbj79) [reply specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json]: @KrisBash - These are already exposed in existing API. Attaching screenshot.

image

File: specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json [hunk 4]
```diff
@@ -0,0 +1,1487 @@
...
+                    },
+                    {
+                        "$ref": "#/parameters/SubscriptionIdParameter"
+                    },
+                    {
+                        "$ref": "#/parameters/ApiVersionParameter"
+                    }
+                ],
+                "responses": {
+                    "200": {
```
[the remark this answers is not in the document]
Comment (shahbj79) [reply specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json]: Yes, this is always synchronous

File: specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json [hunk 5]
```diff
@@ -0,0 +1,1487 @@
...
+                    }
+                },
+                "x-ms-examples": {
+                    "GetCommunicationDetailsForSubscriptionSupportTicket": {
+                        "$ref": "./examples/GetCommunicationDetailsForSubscriptionSupportTicket.json"
+                    }
+                }
+            }
+        },
+        "/subscriptions/{subscriptionId}/providers/Microsoft.Support/supportTickets/{supportTicketName}/addCommunication": {
```
[the remark this answers is not in the document]
Comment (shahbj79) [reply specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json]: Supporting PUT for this resource type will be very tricky - 
1. As communication is stored in MSaaS backend that generates a unique id. 
2. Communication also includes email communications (or phone communications) sent by support engineers to customers via MSaaS backend system and these wouldn't have a name.

File: specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json [hunk 6]
```diff
@@ -0,0 +1,1487 @@
...
+            "description": "The error details."
+        },
+        "CreateSupportTicket": {
+            "description": "Describes the payload for creating a support ticket. Customers can create tickets for issues relating to Billing, Subscription Management, Service and subscription limits(quotas) and Technical as determined by service id in the request payload.",
+            "required": [
+                "properties"
+            ],
+            "type": "object",
+            "properties": {
+                "properties": {
```
[the remark this answers is not in the document]
Comment (shahbj79) [reply specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json]: I have added the common resource properties (id, name, type). However, the request and response is not the same as there are more properties that are being returned in the response. Request definition is CreateSupportTicket and response definition is SupportTicketDetails as there are additional properties such as sla, status, etc being returned in the SupportTicketDetails.

[the remark this answers is not in the document]
Comment (shahbj79) [reply specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json]: Let me know if there is another way to handle this. Should I be using SupportTicketDetails instead and marking all additional properties as readonly?

File: specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json [hunk 7]
```diff
@@ -0,0 +1,1502 @@
...
+                    }
+                },
+                "x-ms-examples": {
+                    "GetIssueClassification": {
+                        "$ref": "./examples/GetProblemClassification.json"
+                    }
+                }
+            }
+        },
+        "/subscriptions/{subscriptionId}/providers/Microsoft.Support/checkNameAvailability": {
```
Comment (anthony-c-martin) [inline specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json]: Why is this API necessary? What's the scope of the name that you're checking for availability - is this something that must be globally unique?

File: specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json [hunk 8]
```diff
@@ -0,0 +1,1502 @@
...
+                "type"
+            ],
+            "properties": {
+                "name": {
+                    "description": "The resource name to validate.",
+                    "type": "string"
+                },
+                "type": {
+                    "type": "string",
+                    "enum": [ "Microsoft.Support/supportTickets" ],
```
Comment (anthony-c-martin) [inline specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json]: Are there plans to add new resource types to this list in the future?

File: specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json (hunk 7 above)
Comment (shahbj79) [reply specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json]: This is based on the feedback from Mark Simms/Gaurav/Kris. I will forward you that thread. The ticket is unique at the subscription scope.

File: specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json (hunk 8 above)
Comment (shahbj79) [reply specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json]: Added one more "Microsoft.Support/communications" to this list. No plans to add more in the near future.

File: specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (shahbj79) [reply specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json]: @KrisBash - removed POST and added PUT for Communication resource as well.

File: specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (shahbj79) [reply specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json]: @KrisBash - I have updated this to be async due to ARM's 2 sec requirement.

File: specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (shahbj79) [reply specification/support/resource-manager/Microsoft.Support/preview/2019-05-01-preview/support.json]: @KrisBash - I have updated request and response definition to be the same.

**Review by anthony-c-martin — APPROVED** [review]

**Review by weidongxu-microsoft — APPROVED** [review]

**Comment by weidongxu-microsoft** [comment]
@KrisBash Please let us know if you are satisfied with Bhavin's update. If you are OK with it, please approve the PR so I can merge it.

**Comment by ryansbenson** [comment]
@KrisBash is on paternity but Anthony signed off. If someone adds 'ARMSignedOff' to your PR, you can consider it approved by ARM all up.

**Comment by weidongxu-microsoft** [comment]
@akning-ms 

Are we able for force merge this?

According to Anthony & Ryan, ARM already signed-off.

Kris is OOF but his change request make me not able to merge the PR. Kris's change request is already addressed, but he probably not have time to approve before his OOF.

**Pull request merged**